File: macos-app/Lexis/Import/PageRangeParser.swift
import Foundation

/// Parses/formats the "1-3,5,7-9" page range syntax used by the PDF page
/// picker. Input is 1-based and inclusive (matching the reference backend);
/// output indices are 0-based, sorted, deduped, and clamped to pageCount.
enum PageRangeParser {
    static func parse(_ text: String, pageCount: Int) -> [Int] {
        guard pageCount > 0 else { return [] }
        var indices = Set<Int>()
        for rawPart in text.split(separator: ",") {
            let part = rawPart.trimmingCharacters(in: .whitespaces)
            guard !part.isEmpty else { continue }
            if let dashIndex = part.firstIndex(of: "-") {
                let startText = part[part.startIndex..<dashIndex].trimmingCharacters(in: .whitespaces)
                let endText = part[part.index(after: dashIndex)...].trimmingCharacters(in: .whitespaces)
                guard let start = Int(startText), let end = Int(endText), start <= end else { continue }
                for page in start...end {
                    let index = page - 1
                    if index >= 0 && index < pageCount { indices.insert(index) }
                }
            } else if let page = Int(part) {
                let index = page - 1
                if index >= 0 && index < pageCount { indices.insert(index) }
            }
        }
        return indices.sorted()
    }

    static func format(_ indices: [Int]) -> String {
        let sorted = indices.sorted()
        guard !sorted.isEmpty else { return "" }

        var groups: [[Int]] = []
        for index in sorted {
            if let lastValue = groups.last?.last, lastValue == index - 1 {
                groups[groups.count - 1].append(index)
            } else {
                groups.append([index])
            }
        }
        return groups.map { group in
            let start = group.first! + 1
            let end = group.last! + 1
            return start == end ? "\(start)" : "\(start)-\(end)"
        }.joined(separator: ",")
    }
}

File: android-app/app/src/main/java/com/aboooooo57/lexume/support/PageRangeParser.kt
package com.aboooooo57.lexume.support

/**
 * Parses/formats the "1-3,5,7-9" page range syntax used by the PDF page
 * picker. Mirrors `Support/PageRangeParser.swift` exactly. Input is 1-based
 * and inclusive; output indices are 0-based, sorted, deduped, and clamped to
 * pageCount.
 */
object PageRangeParser {
    fun parse(text: String, pageCount: Int): List<Int> {
        if (pageCount <= 0) return emptyList()
        val indices = sortedSetOf<Int>()
        for (rawPart in text.split(",")) {
            val part = rawPart.trim()
            if (part.isEmpty()) continue
            val dashIndex = part.indexOf('-')
            if (dashIndex >= 0) {
                val start = part.substring(0, dashIndex).trim().toIntOrNull()
                val end = part.substring(dashIndex + 1).trim().toIntOrNull()
                if (start == null || end == null || start > end) continue
                for (page in start..end) {
                    val index = page - 1
                    if (index in 0 until pageCount) indices.add(index)
                }
            } else {
                val page = part.toIntOrNull() ?: continue
                val index = page - 1
                if (index in 0 until pageCount) indices.add(index)
            }
        }
        return indices.toList()
    }

    fun format(indices: List<Int>): String {
        val sorted = indices.sorted()
        if (sorted.isEmpty()) return ""

        val groups = mutableListOf<MutableList<Int>>()
        for (index in sorted) {
            val lastGroup = groups.lastOrNull()
            if (lastGroup != null && lastGroup.last() == index - 1) {
                lastGroup.add(index)
            } else {
                groups.add(mutableListOf(index))
            }
        }
        return groups.joinToString(",") { group ->
            val start = group.first() + 1
            val end = group.last() + 1
            if (start == end) "$start" else "$start-$end"
        }
    }
}
